### ISSUE TITLE
Architecture: modular, efficient framework for NeRF and 3D Gaussian Splatting research

### ISSUE THREAD

**Issue by hadleyhzy34** [hadleyhzy34/nerf#6]
## Summary

Turn the current educational NeRF implementation into a reusable, efficient research package for implementing and fairly benchmarking multiple Neural Radiance Field (NeRF) and 3D Gaussian Splatting (3DGS) variants.

The current repository is a useful reference baseline, but its model, sampling, rendering, training, evaluation, configuration, and checkpoint logic are tightly coupled. Before adding variants, we should establish small, typed contracts and a benchmark harness so new methods can reuse infrastructure without forcing NeRF and 3DGS into one artificial abstraction.

## Current-state findings

- `src/nerf_project/models/nerf.py` combines positional encoding, the NeRF MLP, PDF sampling, and volume rendering in one module.
- Coarse/fine rendering logic and model reconstruction are duplicated across train, evaluation, and novel-view rendering scripts.
- The Blender loader owns image loading, camera parsing, ray generation, background compositing, caching, and split-specific batching behavior.
- The training dataset returns all rays for one image, while `batch_size` is described as rays per batch; the DataLoader currently batches images rather than sampling a flat ray batch.
- Evaluation references `args.near` and `args.far`, but those arguments are not declared by the evaluation parser.
- Checkpoints serialize an unversioned CLI namespace and script-specific state, making compatibility across variants fragile.
- Tests cannot currently be collected because `tests/test_nerf_model.py` ends with a stray Markdown fence.
- Runtime, evaluation, visualization, logging, and test dependencies are all installed as mandatory package dependencies.
- Package metadata and README paths/names do not consistently match the actual `src/nerf_project` layout.

## Goals

- Make datasets, camera models, scene normalization, metrics, experiment tracking, checkpointing, and benchmark reporting reusable across NeRF and 3DGS.
- Make NeRF components independently replaceable: encoding, field, ray sampler, renderer/integrator, and loss.
- Give 3DGS its own native components: Gaussian parameters, initialization, projection/rasterization adapter, densification/pruning strategy, and losses.
- Support adding a method through composition and registration, with minimal changes to shared engine code.
- Produce comparable quality, training-speed, render-speed, memory, and model-size results with saved provenance.
- Keep the default core lightweight while allowing optional CUDA/rasterizer, perceptual-metric, visualization, and development dependencies.
- Preserve a readable vanilla NeRF as the first reference method.

## Non-goals

- A single universal renderer interface that hides meaningful differences between ray marching and Gaussian rasterization.
- Building a distributed training framework in the first migration.
- Implementing every published NeRF/3DGS variant before the package contracts stabilize.
- Requiring custom CUDA kernels for the reference CPU/PyTorch smoke path.

## Design principles

1. **Shared scene contract, backend-specific rendering.** Cameras, images, splits, normalization, metrics, and experiment records are common; ray and Gaussian renderers remain explicit.
2. **Composition over deep inheritance.** Use small `Protocol`/ABC boundaries and typed dataclasses for data exchanged between components.
3. **Tensor contracts are public API.** Document shape, dtype, device, coordinate convention, value range, and batching semantics for every exchanged tensor.
4. **Config selects components.** Instantiate registered components from structured, serializable configs; do not pass a global CLI namespace through library code.
5. **Scripts are thin entry points.** Training, evaluation, rendering, and benchmarking should call package APIs rather than duplicate pipelines.
6. **Correctness before acceleration.** Keep a deterministic reference path, then add AMP, compilation, fused/custom kernels, and backend-specific acceleration behind measured options.
7. **Benchmarks are reproducible artifacts.** Record commit, resolved config, seed, dataset fingerprint, dependency versions, device, precision, warm-up, and synchronization policy.

## Proposed package structure

The final package name can be confirmed separately; `representation3d` is used below as a neutral working name.

```text
src/representation3d/
├── api/
│   ├── types.py              # Camera, ImageBatch, RayBundle, RenderOutput
│   ├── protocols.py          # Dataset, Representation, Renderer, Metric contracts
│   └── registry.py           # Explicit component/method registration
├── data/
│   ├── base.py
│   ├── blender.py
│   ├── scene.py              # Scene metadata, split definitions, normalization
│   ├── cameras.py            # Intrinsics/extrinsics and convention conversion
│   ├── rays.py               # Ray generation and ray-batch sampling
│   └── transforms.py
├── methods/
│   ├── nerf/
│   │   ├── config.py
│   │   ├── method.py         # Wires field, samplers, renderer, losses
│   │   ├── encodings.py
│   │   ├── fields.py
│   │   ├── samplers.py       # Stratified and hierarchical sampling
│   │   ├── rendering.py      # Volume compositing/integration
│   │   └── losses.py
│   └── gaussian_splatting/
│       ├── config.py
│       ├── method.py
│       ├── model.py          # Means, covariance/scale/rotation, opacity, SH
│       ├── initialization.py
│       ├── rasterization.py  # Backend adapter and reference contract
│       ├── density_control.py
│       └── losses.py
├── engine/
│   ├── trainer.py
│   ├── evaluator.py
│   ├── inference.py
│   ├── checkpoint.py         # Versioned, method-aware checkpoint schema
│   ├── callbacks.py
│   └── reproducibility.py
├── metrics/
│   ├── image.py              # PSNR/SSIM
│   ├── perceptual.py         # Optional LPIPS
│   └── performance.py        # Time, throughput, peak memory, model size
├── benchmark/
│   ├── spec.py               # Dataset/method/budget benchmark definitions
│   ├── runner.py
│   ├── profiler.py
│   └── report.py             # JSON/CSV plus human-readable summary
├── config/
│   ├── schema.py             # Typed dataclass configs
│   └── loading.py
├── cli/
│   ├── train.py
│   ├── evaluate.py
│   ├── render.py
│   └── benchmark.py
└── utils/                    # Only genuinely cross-cutting helpers

configs/
├── methods/
├── datasets/
├── benchmarks/
└── experiments/

tests/
├── unit/
├── integration/
├── regression/
└── benchmarks/               # Non-default performance smoke tests

examples/
├── train_vanilla_nerf.py
├── add_nerf_variant.py
└── benchmark_methods.py
```

## Core contracts

Start with concrete dataclasses and minimal protocols; expand only when a second implementation proves the need.

- `Camera`: batched intrinsics, camera-to-world transform, image size, projection type, convention.
- `SceneBatch`: camera(s), optional RGB/alpha/depth/mask, image/sample IDs, scene metadata.
- `RayBundle`: origins, directions, pixel coordinates, near/far, camera IDs, optional timestamps.
- `RenderOutput`: RGB plus optional depth, opacity, normals, auxiliary losses/stats.
- `SceneDataset`: returns camera/image observations without deciding training ray sampling.
- `Method`: owns trainable representation and exposes `training_step`, `render`, and method state.
- Backend-specific collaborators:
  - NeRF: `Encoding`, `Field`, `RaySampler`, `VolumeRenderer`.
  - 3DGS: `GaussianModel`, `Rasterizer`, `InitializationStrategy`, `DensityControlStrategy`.
- `Metric`: streaming update/compute/reset with declared required outputs.
- `CheckpointSchema`: schema version, method identifier/config, model/optimizer/scheduler/scaler states, step, RNG states, and dataset metadata.

### Phase 0: Stabilize the baseline

- [ ] Remove the test syntax error and make the existing suite collect on supported Python versions.
- [ ] Add a tiny synthetic scene fixture and end-to-end CPU smoke test.
- [ ] Fix ray-batch semantics so `batch_size` means rays, not images containing all rays.
- [ ] Fix train/eval/render argument drift, including evaluation near/far bounds.
- [ ] Add deterministic tests for hierarchical sampling and volume compositing.
- [ ] Capture a vanilla NeRF regression fixture: fixed seed, config, expected shapes, finite outputs, and short-run loss decrease.

### Phase 1: Establish package foundations

- [ ] Confirm package name and correct project metadata.
- [ ] Add typed config, tensor dataclasses, camera conventions, and scene dataset contracts.
- [ ] Split Blender observation loading from ray generation and ray sampling.
- [ ] Split the monolithic NeRF module into encoding, field, sampler, and renderer modules.
- [ ] Move orchestration out of scripts into engine APIs; retain thin CLI wrappers.
- [ ] Introduce a versioned checkpoint format and compatibility loader for current checkpoints where practical.
- [ ] Split dependencies into core and extras such as `eval`, `viz`, `gsplat`, and `dev`.

### Phase 2: Prove NeRF extensibility

- [ ] Register vanilla NeRF as a method assembled from the extracted components.
- [ ] Add one deliberately small variant by swapping a component, such as Fourier encoding vs. identity/hash encoding or uniform vs. occupancy-aware sampling.
- [ ] Verify that the variant requires no changes to dataset, trainer, evaluator, checkpoint, or report code.
- [ ] Add chunked rendering, AMP, optional `torch.compile`, and peak-memory instrumentation behind config flags.
- [ ] Measure output parity and performance against the stabilized baseline.

### Phase 3: Add a native 3DGS path

- [ ] Implement a clear Gaussian parameter/state model and initialization contract.
- [ ] Add a rasterizer adapter so an optimized optional backend can coexist with a testable reference/mock path.
- [ ] Implement optimizer parameter groups and pluggable densification/pruning.
- [ ] Reuse cameras, scenes, metrics, experiment metadata, checkpoints, CLI concepts, and benchmark reporting.
- [ ] Do not route 3DGS through `RayBundle` or the NeRF volume renderer.

### Phase 4: Benchmark harness and CI

- [ ] Add benchmark specs with fixed dataset split, image resolution, training budget, seeds, evaluation views, and precision.
- [ ] Report PSNR, SSIM, optional LPIPS, train time, iterations/sec, rays/sec or splats/sec, render FPS, peak allocated memory, checkpoint size, and parameter/Gaussian count.
- [ ] Include warm-up iterations and device synchronization for accelerator timing.
- [ ] Save JSON and CSV results plus resolved config and environment metadata.
- [ ] Add fast CPU correctness CI and opt-in scheduled GPU regression/performance jobs.
- [ ] Define regression thresholds separately from exploratory benchmark results.

## Benchmark fairness rules

- Compare methods at explicit budgets: wall-clock, iteration count, and/or target quality; never mix them silently.
- Use identical dataset preprocessing, camera convention, train/eval splits, resolution, and metric implementation.
- Run multiple seeds for quality claims and report mean plus spread.
- Separate training throughput from rendering throughput.
- Synchronize CUDA around timed regions and report warm-up policy, precision, device, and backend.
- Track both final quality and quality-over-time so a fast-converging method is visible.
- Treat performance numbers as hardware-specific artifacts, not universal constants.

## Acceptance criteria

- [ ] `pip install .` installs a lightweight core; optional features install through documented extras.
- [ ] Vanilla NeRF can be trained, evaluated, rendered, resumed, and benchmarked through package APIs and CLI entry points.
- [ ] Existing baseline behavior is covered by deterministic regression tests, with any intentional numerical changes documented.
- [ ] A new NeRF variant can be added by implementing/registering method components without editing the shared trainer or evaluator.
- [ ] A minimal 3DGS method reuses shared scene/camera/metric/experiment infrastructure while retaining its own rasterization and density-control path.
- [ ] One command can run a benchmark matrix and emit machine-readable results with full provenance.
- [ ] CPU smoke tests run in CI; accelerator-specific tests skip cleanly when unavailable.
- [ ] Public APIs document tensor shapes, device/dtype expectations, coordinate conventions, and stability status.
- [ ] No train/eval/render pipeline logic is duplicated in top-level scripts.

## Suggested follow-up issues

Once this architecture is agreed, split implementation into focused issues:

1. Baseline correctness and regression fixture.
2. Scene/camera/ray data contracts.
3. NeRF component extraction.
4. Shared engine and versioned checkpoints.
5. Typed configuration and registry.
6. Benchmark runner and result schema.
7. NeRF variant extensibility proof.
8. 3DGS model and rasterizer adapter.
9. Packaging, optional dependencies, CI, and documentation.

## Open decisions

- Final distribution/import name.
- Minimum supported Python/PyTorch/CUDA versions.
- Config library: standard dataclasses plus YAML/TOML loader, or a small established structured-config dependency.
- First optimized 3DGS rasterizer backend and its license/platform constraints.
- Initial benchmark datasets and resource budgets.

**Comment by hadleyhzy34** [comment]
Implementation issue index after Phase 0 completion:

- #7 — Phase 0 stabilization (**completed**)
- #9 — Phase 1A package identity and source layout
- #10 — Phase 1B shared scene/camera/observation/ray contracts
- #11 — Phase 1C modular vanilla NeRF extraction
- #12 — Phase 1D shared engine, typed config, registry, and checkpoints
- #13 — Phase 2A benchmark runner, profiling, efficiency switches, and packaging
- #8 — Phase 2A versioned small-dataset/basic-NeRF demo
- #14 — Phase 2B NeRF extensibility proof
- #15 — Phase 3 native 3D Gaussian Splatting path

Dependency order:

`#7 → #9 → #10 → #11 → #12 → #13 → #8 → #14 → #15`

Issue #8 was rewritten to consume the shared APIs and benchmark schema rather than extending the current scripts. Benchmark infrastructure intentionally precedes the published demo and 3DGS work.
